### PR TITLE
Replace `IterableSubprocessError` with `CommandError`

### PR DESCRIPTION
**Pull request by christian-monch**
This PR implements the solution agreed upon in issue #567, comment https://github.com/datalad/datalad-next/issues/567#issuecomment-1860313394

It replaces `IterableSubprocessError` with `CommandError` to increase compatibility of `iter_subproc` (and `iterable_subprocesses`) with datalad-core.

### CODE REVIEW

**Comment by christian-monch** [comment]
There seems to be some problem with retrieving the docker image for the sshd. This is an excerpt of the log of the `linux`-build.

```
[00:01:38] Running "before_test" scripts
[00:01:38] [ "$DATALAD_TESTS_SSH" = 1 ] && tools/appveyor/setup-sshd || true
[00:01:38] + DATALAD_TESTS_SERVER_SSH_SECKEY=/home/appveyor/.ssh/datalad_tests_id_rsa
[00:01:38] + docker -v
[00:01:38] Docker version 24.0.5, build ced0996
[00:01:38] + setup_docker
[00:01:38] + curl -fsSL --ssl-no-revoke -o sshd.dockerimg.gz 'https://ci.appveyor.com/api/projects/mih/datalad-ci-docker-containers/artifacts/sshd.dockerimg.gz?job=sshd'
[00:01:38] curl: (22) The requested URL returned error: 404 
[00:01:38] + gzip -c -d sshd.dockerimg.gz
[00:01:38] + docker load
[00:01:38] gzip: sshd.dockerimg.gz: No such file or directory
```

**Comment by adswa** [comment]
I think this docker image is generated in this Appveyor project: https://ci.appveyor.com/project/mih/datalad-ci-docker-containers. I believe the artifacts are only kept for 3 months, which is when this job last ran. I rebuild the sshd job, and will rerun the remaining CI in this PR afterwards.

**Comment by adswa** [comment]
there is a single Windows error:

```
_________________________ test_force_checkdatapresent _________________________
srcpath = 'C:\\DLTMP\\datalad_temp_y_hwu8md'
dstpath = 'C:\\DLTMP\\datalad_temp_e77l24wi'
    @with_tempfile(mkdir=True)
    @with_tempfile(mkdir=True)
    def test_force_checkdatapresent(srcpath=None, dstpath=None):
        src = Dataset(srcpath).create()
        target = mk_push_target(src, 'target', dstpath, annex=True, bare=True)
        (src.pathobj / 'test_mod_annex_file').write_text("Heavy stuff.")
        src.save(to_git=False, message="New annex file")
        assert_repo_status(src.path, annex=True)
        whereis_prior = src.repo.whereis(files=['test_mod_annex_file'])[0]
    
        res = src.push(to='target', data='nothing')
        # nothing reported to be copied
        assert_not_in_results(res, action='copy')
        # we got the git-push nevertheless
        eq_(src.repo.get_hexsha(DEFAULT_BRANCH), target.get_hexsha(DEFAULT_BRANCH))
        # nothing moved
        eq_(whereis_prior, src.repo.whereis(files=['test_mod_annex_file'])[0])
    
        # now a push without forced no-transfer
        # we do not give since, so the non-transfered file is picked up
        # and transferred
        annex_branch_state = src.repo.get_hexsha('git-annex')
        res = src.push(to='target', force=None)
        # no branch change, done before
        assert_in_results(res, action='publish', status='notneeded',
                          refspec=DEFAULT_REFSPEC)
        # but availability update
        # reflected as a change in git-annex branch
        assert annex_branch_state != src.repo.get_hexsha('git-annex')
        # and identical state here and on the remote
        assert src.repo.get_hexsha('git-annex') == src.repo.get_hexsha('target/git-annex')
    
        # as a result of [10.20231129-84-geb59da9dd2](https://git.kitenet.net/index.cgi/git-annex.git/commit/?id=HEAD)
        # which makes git-annex clocks less precise, the git-annex branch change on the remote as a result of
        # `annex copy` has exactly the same content and timestamp in git-annex branch, so the same commit as local,
        # so when we fetch updates from the remote, we end up on the same commit instead of merging
        # updated availability with difference in timestamp within git-annex branch of minuscule difference
        try:
>           assert_in_results(res, action='publish', status='notneeded',
                              refspec='refs/heads/git-annex:refs/heads/git-annex')
C:\Python39-x64\lib\site-packages\datalad\core\distributed\tests\test_push.py:496: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Python39-x64\lib\site-packages\datalad\tests\utils_pytest.py:1396: in assert_in_results
    _check_results_in(True, results, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
should_contain = True
results = [{'action': 'copy', 'annexkey': 'MD5E-s12--f01b6f1223d1baf9cc6288f9172bbbf5', 'message': 'to target...', 'path': 'C:\\... 'message': 'refs/heads/git-annex->target:refs/heads/git-annex 03ec22e..8fde746', 'operations': ['fast-forward'], ...}]
kwargs = {'action': 'publish', 'refspec': 'refs/heads/git-annex:refs/heads/git-annex', 'status': 'notneeded'}
found = False, msg = 'Desired result\n{}\nnot found among\n{}'
    def _check_results_in(should_contain, results, **kwargs):
        results = ensure_result_list(results)
        found = False
        for r in results:
            if all(k in r and r[k] == v for k, v in kwargs.items()):
                found = True
                break
        if found ^ should_contain:
            if should_contain:
                msg = "Desired result\n{}\nnot found among\n{}"
            else:
                msg = "Result\n{}\nunexpectedly found among\n{}"
>           raise AssertionError(msg.format(_format_res(kwargs),
                                            _format_res(results)))
E           AssertionError: Desired result
E             {
E              "action": "publish",
E              "refspec": "refs/heads/git-annex:refs/heads/git-annex",
E              "status": "notneeded"
E             }
E           not found among
E             [
E              {
E               "action": "copy",
E               "annexkey": "MD5E-s12--f01b6f1223d1baf9cc6288f9172bbbf5",
E               "message": "to target...",
E               "path": "C:\\DLTMP\\datalad_temp_y_hwu8md\\test_mod_annex_file",
E               "refds": "C:\\DLTMP\\datalad_temp_y_hwu8md",
E               "status": "ok",
E               "target": "target",
E               "type": "file"
E              },
E              {
E               "action": "publish",
E               "hints": null,
E               "message": "refs/heads/dl-test-branch->target:refs/heads/dl-test-branch [up to date]",
E               "operations": [
E                "uptodate"
E               ],
E               "path": "C:\\DLTMP\\datalad_temp_y_hwu8md",
E               "refds": "C:\\DLTMP\\datalad_temp_y_hwu8md",
E               "refspec": "refs/heads/dl-test-branch:refs/heads/dl-test-branch",
E               "status": "notneeded",
E               "target": "target",
E               "type": "dataset"
E              },
E              {
E               "action": "publish",
E               "hints": null,
E               "message": "refs/heads/git-annex->target:refs/heads/git-annex 03ec22e..8fde746",
E               "operations": [
E                "fast-forward"
E               ],
E               "path": "C:\\DLTMP\\datalad_temp_y_hwu8md",
E               "refds": "C:\\DLTMP\\datalad_temp_y_hwu8md",
E               "refspec": "refs/heads/git-annex:refs/heads/git-annex",
E               "status": "ok",
E               "target": "target",
E               "type": "dataset"
E              }
E             ]
C:\Python39-x64\lib\site-packages\datalad\tests\utils_pytest.py:1389: AssertionError
```

**Comment by christian-monch** [comment]
> I think this docker image is generated in this Appveyor project: https://ci.appveyor.com/project/mih/datalad-ci-docker-containers. I believe the artifacts are only kept for 3 months, which is when this job last ran. I rebuild the sshd job, and will rerun the remaining CI in this PR afterwards.

Thanks a lot @adswa

**Comment by christian-monch** [comment]
> there is a single Windows error that seems unrelated:
> 

Thanks for the pointer, looking into it.

**Comment by mih** [comment]
> I believe the artifacts are only kept for 3 months, which is when this job last ran.

I have now adjusted the setup to have the build run once a month on a fixed schedule that does not require manual intervention.

Thanks for spotting the issue!

**Review by mih — CHANGES_REQUESTED** [review]
Thanks for approaching this issue!

I think your solution is technically sound. Yet, I want to propose a modification that hopefully reduces the tension between the `subprocess` world and `datalad-core` a bit more.

1. Let's move `datalad_next/runners/exceptions.py` underneath `datalad_next/exceptions`, maybe as `commanderror.py`. `CommandError` is already imported there (too). This location would not have the circular import problem. We can add a back-import to `datalad_next/runners/exceptions.py` also, with a note that this is legacy. It should remain imported in ``datalad_next/runners/__init__.py` (but from the new location).
2. Let's add `returncode` as an alias to `code` to `CommandError`. Anyone familiar with `subprocess` (many) expects it that way. This avoids most of the diff in this PR, and also minimized the changes to `iterable_subprocess.py` -- which ideally stays as unmodified as anyhow possible.

**Comment by christian-monch** [comment]
> Thanks for approaching this issue!
> 
> I think your solution is technically sound. Yet, I want to propose a modification that hopefully reduces the tension between the `subprocess` world and `datalad-core` a bit more.
> 
> 1. Let's move `datalad_next/runners/exceptions.py` underneath `datalad_next/exceptions`, maybe as `commanderror.py`. `CommandError` is already imported there (too). This location would not have the circular import problem. We can add a back-import to `datalad_next/runners/exceptions.py` also, with a note that this is legacy. It should remain imported in ``datalad_next/runners/**init**.py` (but from the new location).

I will do that.

> 2. Let's add `returncode` as an alias to `code` to `CommandError`. Anyone familiar with `subprocess` (many) expects it that way. This avoids most of the diff in this PR, and also minimized the changes to `iterable_subprocess.py` -- which ideally stays as unmodified as anyhow possible.

Great idea, I was feeling uneasy with the different names from the two "worlds". Will do that.

**Comment by mih** [comment]
Can you fix the conflict, so the test run? Thanks!

**Comment by christian-monch** [comment]
The current approach is: 

1. Use the redefined `datalad_next.runners.exceptions.CommandError` to signal errors from `iterable_subprocess`,
2. Extend `datalad_next.runners.exceptions.CommandError` with the alias `returncode` for `code`,
3. Move `datalad_next.runners.exceptions.CommandError` to `datalad_next.exceptions.commanderror.CommandError`,
4. Import  `datalad_next.runners.exceptions.CommandError` in `datalad_next.runners.__init__.py`.

Point 4. requires a number of additional changes because previously `datalad.runner.exception.CommandError` was imported in `datalad_next.runners.__init__.py`. All code-parts that import `CommandError` from `datalad_next.runners` now get the class `datalad_next.exceptions.commanderror.CommandError` instead of `datalad.runner.exception.CommandError`. The former is a subclass of the latter.

All datalad-core code raises `datalad.runner.exception.CommandError`. Code in `datalad_next` that tries to catch that error must now import `datalad.runner.exception.CommandError` instead of `datalad_next.runners.CommandError`. If it uses the later in an except clause, the clause will not be triggered by  `datalad.runner.exception.CommandError`, because this is a superclass of `datalad_next.runners.CommandError`. 

I have done the necessary changes, but I think the solution is suboptimal and the name overloading of `CommandError` is a recipe for desaster. In my opinion it would be better to do the following:

- rename `datalad_next.exceptions.commanderror.CommandError` to something like `datalad_next.exceptions.iterablecommanderror.IterableCommandError`
- go back to importing `datalad.runner.exception.CommandError` into `datalad.runner.exception`
- optional: import `IterableCommandError` into `datalad.runner.exception`.

**Comment by mih** [comment]
> Another possibility would be to just use datalad.runner.exception.CommandError to signal errors in iterable_subprocess. Either as it is, or patched to contain the improved __repr__ and the returncode-alias.

I like this much better TBH. I would still have all code import from `datalad_next.exceptions` and have `datalad.runner.exception.CommandError` be imported there. We can use the standard patch facilities to inject the missing functionality in -core.

**Comment by christian-monch** [comment]
> > Another possibility would be to just use datalad.runner.exception.CommandError to signal errors in iterable_subprocess. Either as it is, or patched to contain the improved **repr** and the returncode-alias.
> 
> I like this much better TBH. I would still have all code import from `datalad_next.exceptions` and have `datalad.runner.exception.CommandError` be imported there. We can use the standard patch facilities to inject the missing functionality in -core.

Will change the PR to this strategy.

File: datalad_next/iterable_subprocess/iterable_subprocess.py [hunk 1]
```diff
@@ -4,6 +4,9 @@
 from subprocess import PIPE, SubprocessError, Popen
 from threading import Thread
 
+# Importing from datalad-core to prevent circular imports
+from datalad.runner.exception import CommandError
```
[the remark this answers is not in the document]
Comment (christian-monch) [reply datalad_next/iterable_subprocess/iterable_subprocess.py]: `CommandError` improvements are now available in `datalad` and `datalad-next` by patching `datalad.runner.exception.CommandError`.

Marking this conversation as resolved.

**Review by mih — APPROVED** [review]
This works for me. Thx!

**Pull request merged**